### PR TITLE
Optionally add ref to git commit with release version 

### PR DESCRIPTION
**Pull request by klutchell**
Use the balena-sdk to get the release `raw_version` after a successful push.

If `create_ref` is true then use octokit to create a ref on the workflow sha.

Change-type: minor
Signed-off-by: Kyle Harding <kyle@balena.io>
Resolves: https://github.com/balena-io/balena-ci/issues/17

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -37,6 +37,7 @@ Inputs are provided using the `with:` section of your workflow YML file.
 | fleet | The slug of the fleet (eg: `my_org/sample_fleet`) for which the release is for | true | |
 | environment | Domain of API hosting your fleets | false | balena-cloud.com |
 | versionbot | Tells action to use Versionbot branch for versioning | false | false |
+| create_ref | Attempt to tag the git commit with the release version | false | false |
```
Comment (20k-ultra) [inline README.md:40]:
```suggestion
| create_ref | Tag the git commit with the release version | false | false |
```

Attempt sounds like it might not tag and complete still.

What do you think about changing the name to just `tag` ?

File: README.md [hunk 2]
```diff
@@ -45,6 +46,7 @@ Inputs are provided using the `with:` section of your workflow YML file.
 | key | Description | Nullable |
 | --- | --- | --- |
 | release_id | ID of the release built | true |
+| raw_version | Version of the release built | true |
```
Comment (20k-ultra) [inline README.md:49]: I know this value is called `raw_version` in the SDK but what do you think about just calling it `version` ? The raw part implies there is another version.

File: src/action.ts [hunk 3]
```diff
@@ -5,6 +5,8 @@ import * as versionbot from './versionbot-utils';
 import * as balena from './balena-utils';
 import * as git from './git';
 
+import { createRef } from './github-utils';
```
Comment (20k-ultra) [inline src/action.ts:8]: Group the local imports

```suggestion
import * as git from './git';
import { createRef } from './github-utils';
```

File: src/action.ts [hunk 4]
```diff
@@ -28,6 +32,17 @@ export async function run(): Promise<void> {
 		});
 		// Set the built releaseId in the output
 		core.setOutput('release_id', releaseId);
+
+		if (releaseId) {
+			rawVersion = await balena.getReleaseVersion(parseInt(releaseId, 10));
+		}
```
Comment (20k-ultra) [inline src/action.ts:38]: Don't have to check if the `releaseId` is set because the push command would have thrown an exception if it didn't

File: src/action.ts [hunk 5]
```diff
@@ -28,6 +32,17 @@ export async function run(): Promise<void> {
...
+
+		core.setOutput('raw_version', rawVersion);
+
+		if (rawVersion && core.getBooleanInput('create_ref', { required: false })) {
```
Comment (20k-ultra) [inline src/action.ts:42]: if we make getReleaseVersion throw when it fails to get the rawVersion then we don't have to check if it's set (I'm going to write a comment about this function that it should throw)

```suggestion
		if (core.getBooleanInput('create_ref', { required: false })) {
```

File: src/action.ts [hunk 6]
```diff
@@ -84,5 +96,15 @@ export async function run(): Promise<void> {
...
 
+	if (releaseId) {
+		rawVersion = await balena.getReleaseVersion(parseInt(releaseId, 10));
+	}
+
+	core.setOutput('raw_version', rawVersion);
+
+	if (rawVersion && core.getBooleanInput('create_ref', { required: false })) {
+		await createRef(rawVersion, context.payload.pull_request?.head.sha);
+	}
```
Comment (20k-ultra) [inline src/action.ts:107]: This duplicate code is making me realize we should move the code between line 56-75 to the beginning since it doesn't try to build a release. Then, we can just build a release and do this createRef code once...we can do it after this PR though

File: src/balena-utils.ts [hunk 7]
```diff
@@ -146,6 +146,29 @@ export async function getReleaseByTags(
 	};
 }
 
+// https://www.balena.io/docs/reference/sdk/node-sdk/#balena.models.release.get
+export async function getReleaseVersion(releaseId: number): Promise<string> {
+	core.debug(`Getting raw version for release ID ${releaseId}`);
+
+	await balena.auth.loginWithToken(
+		core.getInput('balena_token', { required: true }),
+	);
```
Comment (20k-ultra) [inline src/balena-utils.ts:155]: this makes me realize we should think of a better way to authenticate the SDK once...we can do it in another PR

File: src/balena-utils.ts [hunk 8]
```diff
@@ -146,6 +146,29 @@ export async function getReleaseByTags(
...
+
+	return await balena.models.release
+		.get(releaseId, { $select: 'raw_version' })
+		.then((release) => {
```
Comment (20k-ultra) [inline src/balena-utils.ts:159]: I don't like that we are mixing await with then because it's 2 different styles of handling promises so we should just use 1 (await btw).

File: src/balena-utils.ts [hunk 9]
```diff
@@ -146,6 +146,29 @@ export async function getReleaseByTags(
...
+			core.debug(JSON.stringify(release));
+			if (release.raw_version) {
```
Comment (20k-ultra) [inline src/balena-utils.ts:161]: It's not possible for a release to not have a version so not need to check

File: src/balena-utils.ts [hunk 10]
```diff
@@ -146,6 +146,29 @@ export async function getReleaseByTags(
...
+				return release.raw_version;
+			}
+			return '';
```
Comment (20k-ultra) [inline src/balena-utils.ts:164]: I would rather throw an error then return empty string so the action doesn't continue

File: src/balena-utils.ts [hunk 11]
```diff
@@ -146,6 +146,29 @@ export async function getReleaseByTags(
...
+		})
+		.catch((e) => {
+			core.error(e.message);
+			throw new Error(e);
+		});
```
Comment (20k-ultra) [inline src/balena-utils.ts:169]: No need to catch the error just to log and throw it again because the main.ts file catches all exceptions.

File: src/github-utils.ts [hunk 12]
```diff
@@ -37,3 +37,28 @@ export async function getChecks(): Promise<CheckRun[]> {
...
+		})
+		.then((response) => {
+			core.debug(`Created ${response.data.url}`);
+			return response.data.url;
+		})
+		.catch((e) => {
+			core.error(e.message);
+			// TODO: catch error if not 'ref already exists'
+			return '';
+		});
```
Comment (20k-ultra) [inline src/github-utils.ts:63]: Same comments as the `getReleaseVersion` function. However, Here is how I would do both functions so use this example as guide for the release version function.

```suggestion
	const response = await octokit.rest.git
		.createRef({
			owner: github.context.repo.owner,
			repo: github.context.repo.repo,
			ref: `refs/tags/${tag}`,
			sha,
		});
	// TODO: catch error if not 'ref already exists'	
	core.debug(`Created ${response.data.url}`);	
	return response.data.url;	
```

Only catch exception if you will be adding some handling (such as dealing with ref already exists) or want to log a better message.

**Review by 20k-ultra — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (klutchell) [reply README.md:40]: I said "attempt" because we want to ignore failures if the tag already exists. Like pushing to master where versionbot has already created a tag, and the balena dashboard version might be at rev0 (rev0 is not appended) so they would conflict.

That said, we don't need to advertise that we are ignoring certain errors. I agree we could just change this to your suggestion.

File: README.md (hunk 2 above)
Comment (klutchell) [reply README.md:49]: Sure, works for me. I'll take the change.

File: src/action.ts (hunk 3 above)
Comment (klutchell) [reply src/action.ts:8]: Sure thing!

File: README.md (hunk 1 above)
Comment (klutchell) [reply README.md:40]: I think `tag` is already an overloaded word in a lot of contexts, but ultimately it's up to you?

Comment (20k-ultra) [reply README.md]: `create_ref` seems good to me then if the upstream function is called that!

Let's just change the description to not include "Attempt". My thought process is, for debugging I want the action to always do what it's told and error if the state hasn't changed given the instructions provided (if change was expected).

Comment (20k-ultra) [reply README.md]: Feel free to self-certify after reading my comment^

**Comment by 20k-ultra** [comment]
@balena-ci rebase

**Comment by klutchell** [comment]
@balena-ci I self-certify!
